org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java

```diff
@@ -88,16 +88,20 @@ public Message read(JsonReader in) throws IOException {
 					break;
 				}
 				case "params": {
-					Type type = null;
+					Type parameterType = null;
 					if (method != null) {
 						JsonRpcMethod jsonRpcMethod = handler.getJsonRpcMethod(method);
 						if (jsonRpcMethod != null)
-							type = jsonRpcMethod.getParameterType();
+							parameterType = jsonRpcMethod.getParameterType();
 					}
-					if (type == null)
+					if (parameterType == null || parameterType == Void.class) {
 						params = new JsonParser().parse(in);
-					else
-						params = gson.fromJson(in, type);
+					} else {
+						params = gson.fromJson(in, parameterType);
+						if (params instanceof JsonElement) {
+							params = gson.fromJson((JsonElement) params, parameterType);
+						}
+					}
```

Comment:
The reason I had deferred the parsing of JsonElement to `createMessage(..)` is that I'd like to support an arbitrary order of message attributes. This is relevant when `params` appears before `method`. I vote for restoring the previous behavior, except for the special handling of `Void.class`.